cupyx/scipy/signal/signaltools.py

```diff
@@ -181,27 +184,91 @@ def fftconvolve(in1, in2, mode='full', axes=None):
     return _st_core._apply_conv_mode(out, in1.shape, in2.shape, mode, axes)
 
 
-def choose_conv_method(in1, in2, mode='full'):
+def choose_conv_method(in1, in2, mode='full', measure=False):
     """Find the fastest convolution/correlation method.
 
+    This supports timing the convolution to adapt the value of ``method`` to a
+    particular set of inputs and/or hardware.
+
     Args:
         in1 (cupy.ndarray): first input.
         in2 (cupy.ndarray): second input.
         mode (str, optional): ``valid``, ``same``, ``full``.
+        measure (bool, optional): If ``True``, run and time the convolution of
+            ``in1`` and ``in2`` with both methods and return the fastest. If
+            ``False`` (default), predict the fastest method using precomputed
+            values.
 
     Returns:
         str: A string indicating which convolution method is fastest,
-        either ``direct`` or ``fft1``.
+            either ``direct`` or ``fft1``.
+        dict: A dictionary containing the times (in seconds) needed for each
+            method. This is only returned if ``measure=True``.
 
     .. warning::
-        This function currently doesn't support measure option,
-        nor multidimensional inputs. It does not guarantee
-        the compatibility of the return value to SciPy's one.
+        This function does not guarantee the compatibility of the return value
+        to SciPy's one.
 
     .. seealso:: :func:`scipy.signal.choose_conv_method`
-
+    .. seealso:: :func:`cupyx.scipy.signal.convolve`
+    .. seealso:: :func:`cupyx.scipy.signal.correlate`
     """
-    return cupy._math.misc._choose_conv_method(in1, in2, mode)
+    if measure:
+        # Actually measure the times
+        times = {}
+        for method in ('fft', 'direct'):
+            def _timeit(repeats):
+                for _ in range(repeats):
+                    convolve(in1, in2, mode, method)
+            _timeit(1)
+            for p in range(0, 10):
+                n = 10**p
+                time = cupyx.time.repeat(
+                    _timeit, (n,), 1, n_warmup=0).cpu_times.min()
+                if time >= 5e-3 / 10:
+                    break
+            if time < 1:  # second
+                time = cupyx.time.repeat(
+                    _timeit, (n*10,), 3, n_warmup=1).cpu_times.min()
+            times[method] = time
+        chosen_method = 'fft' if times['fft'] < times['direct'] else 'direct'
+        return chosen_method, times
+
+    # Direct method doesn't support smaller array having more than 2 GiB items
+    if min(in1.size, in2.size) >= (1 << 31):
+        return 'fft'
+
+    # For integer input, catch when more precision required than float provides
+    # (representing an integer as float loses precision if larger than 2**52)
+    if in1.dtype.kind in 'ui' or in2.dtype.kind in 'ui':
+        max_value = abs(in1).max() * abs(in2).max() * min(in1.size, in2.size)
+        if max_value > 2**cupy.finfo('float').nmant - 1:
```

Comment:
These lines require synchronizations.

Reply:
IMO, we can always use `direct` mode for integer inputs, because `direct` mode tend to be faster if `min(in1.size, in2.size)` is small.

Reply:
But it could also be needed for int32, uint32, int64, and uint64 even with small sizes (in fact, it is always needed for those). To avoid synchronizations it could use `cupy.iinfo(in1.dtype).max * cupy.iinfo(in2.dtype).max * min(in1.size, in2.size)` or similar. This will miss the situations when the arrays don't use the full range of their values, but avoiding the syncs is good.

Reply:
Actually, maybe you are right. I had it backwards in my mind. Small ones will end up using direct. Large ones will use direct (due to loss of precision using fft). Medium ones should maybe use fft but possibly not so many in that region. If it is assumed that integers are always direct, that makes the compute-personal-values much, much faster - 2/3 of the data types are integers (or bool).